File: Day 31 - Dynamic Programming (Part 1)/Min steps to reach one.java
import java.util.Scanner;

public class Main {
  public static void main(String[] args) {
    Scanner sc = new Scanner(System.in);
    int n = sc.nextInt();
    sc.close();

    int[] dp = new int[n + 1];
    dp[1] = 0;

    for (int i = 2; i <= n; i++) {
      dp[i] = Integer.MAX_VALUE;
      dp[i] = Math.min(dp[i - 1], dp[i]); // N -> N-1
      if (i % 2 == 0) {
        dp[i] = Math.min(dp[i / 2], dp[i]); // N -> N/2
      }
      if (i % 3 == 0) {
        dp[i] = Math.min(dp[i / 3], dp[i]); // N -> N/3
      }
      dp[i]++; // +1 for current operation
    }

    System.out.println(dp[n]);
  }
}

File: Day 31, 32 - Dynamic Programming (Part 1)/Min steps to reach one.cpp
#include <bits/stdc++.h>
using namespace std;

int main()
{
  int n;
  cin >> n;

  int dp[n + 1];
  dp[1] = 0;
  for (int i = 2; i <= n; i++)
  {
    dp[i] = INT_MAX;
    dp[i] = min(dp[i - 1], dp[i]); // N -> N-1
    if (i % 2 == 0)
      dp[i] = min(dp[i / 2], dp[i]); // N -> N/2
    if (i % 3 == 0)
      dp[i] = min(dp[i / 3], dp[i]); // N -> N/3

    dp[i]++; // +1 for current operation
  }
  cout << dp[n];

  return 0;
}